job-server-extras/src/spark.jobserver/python/PythonJob.scala

```diff
@@ -0,0 +1,148 @@
+package spark.jobserver.python
+
+import com.typesafe.config.Config
+import org.scalactic.{Good, Every, Or}
+import org.slf4j.LoggerFactory
+import py4j.GatewayServer
+import spark.jobserver.api.{SparkJobBase, ValidationProblem, JobEnvironment}
+
+import scala.sys.process.{ProcessLogger, Process}
+import scala.util.{Failure, Success, Try}
+import scala.collection.JavaConverters._
+
+object PythonJob {
+
+  case class ConversionException(unconvertibleValue: Any, message: String) extends Exception(message)
+
+  /**
+    * Py4J will autoconvert the jobOutput from Python primitives and collections
+    * to Java primitives and collections from java.util see
+    * https://www.py4j.org/advanced_topics.html#converting-python-collections-to-java-collections
+    *
+    * This method converts such objects to Scala types that we expect spray.json
+    * to be able to serialize.
+    *
+    * Due to type erasure, we cannot check that collections are collections of a type
+    * that can subsequently be serialized so this is not a fool-proof check.
+    *
+    * @param rawResult the object returned from the Python subprocess.
+    * @return the Scala equivalent if successful, else an error.
+    */
+  def convertRawResult(rawResult: Any): Try[Any] = rawResult match {
+    case z: Boolean => Success(z)
+    case b: Byte => Success(b)
+    case c: Char => Success(c)
+    case s: Short => Success(s)
+    case i: Int => Success(i)
+    case l: Long => Success(l)
+    case f: Float => Success(f)
+    case d: Double => Success(d)
+    case s: String => Success(s)
+    case m: java.util.HashMap[_,_] =>
+      val scalaMap = m.asScala.toMap
+      val recursedMap = scalaMap.mapValues(convertRawResult)
+      recursedMap.foldLeft(Success(Map.empty[Any, Any]): Try[Map[Any, Any]]) {
```

Comment:
Neat, recursive functional validation of conversion.   There must be some library that does this already, but I can't think of anything off the top of my head.   

This might be expensive if the result set is fairly large.  I'd be OK with just a dumb try/catch solution too.